gdal/port/cpl_json.h

```diff
@@ -0,0 +1,216 @@
+/******************************************************************************
+ * Project:  Common Portability Library
+ * Purpose:  Function wrapper for libjson-c access.
+ * Author:   Dmitry Baryshnikov, dmitry.baryshnikov@nextgis.com
+ *
+ ******************************************************************************
+ * Copyright (c) 2016-2017 NextGIS, <info@nextgis.com>
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ ****************************************************************************/
+
+#ifndef CPL_JSON_H_INCLUDED
+#define CPL_JSON_H_INCLUDED
+
+#include "cpl_string.h"
+#include "cpl_progress.h"
+
+#include <vector>
+
+/**
+ * \file cpl_json.h
+ *
+ * Interface for read and write JSON documents
+ */
+
+
+
+/*! @cond Doxygen_Suppress */
+typedef void *JSONObjectH;
+
+CPL_C_START
+
+class CPLJSONArray;
+/*! @endcond */
+
+/**
+ * @brief The CPLJSONArray class holds JSON object from CPLJSONDocument
+ */
+class CPL_DLL CPLJSONObject
+{
+    friend class CPLJSONArray;
+    friend class CPLJSONDocument;
+public:
+    /**
+     * Json object types
+     */
+    enum Type {
+        Unknown,
+        Null,
+        Object,
+        Array,
+        Boolean,
+        String,
+        Integer,
+        Long,
+        Double
+    };
+
+    /**
+     * Json object format to string options
+     */
+    enum PrettyFormat {
+        Plain,  ///< No extra whitespace or formatting applied
+        Spaced, ///< Minimal whitespace inserted
+        Pretty  ///< Formated output
+    };
+
+public:
+/*! @cond Doxygen_Suppress */
+    CPLJSONObject();
+    explicit CPLJSONObject(const char *pszName, const CPLJSONObject &oParent);
+    ~CPLJSONObject();
+    CPLJSONObject(const CPLJSONObject &other);
+    CPLJSONObject &operator=(const CPLJSONObject &other);
+
+#if !_MSC_VER
```

Comment:
why is this exception for MSC_VER necessary ?

Reply:
There was strange feature/bug in VS2013. Now this is not necessary, I'll remove this define.